Implement the DS Reputation System.

nnamon opened the pull request:
## Description
<!-- What is the overall goal of your pull request? -->
<!-- What is the context of your pull request? -->
<!-- What are the related issues and pull requests? -->

- Implements the DS Reputation proposal as written here: https://github.com/nnamon/zilliqa-research/blob/master/ds_reputation/proposal.md.
- Diverges from the proposal by adding a separate constant for the max number of Byzantine nodes to remove to allow for node turnover eventuality.
- Diverges from the proposal by deriving the performance from the coinbase rewards rather than tracking it separately.
- Add two new constants: 
  - `DS_PERFORMANCE_THRESHOLD_PERCENT`: the ratio for the minimum number of co-sigs for a DS node to be considered non-Byzantine (0.0 - 1.0).
  - `NUM_DS_BYZANTINE_REMOVED`: the number of Byzantine DS nodes to remove. Must be less than or equal to `NUM_DS_ELECTION`.
- Diverges from the proposal to use the powWinners field and add a new field to the DS Block Header structure: `m_removeDSNodePubkeys` 
  - Update the protobuf definition.
  - Update the Messenger to serialize and deserialize the new field.
  - Increment `DSBLOCK_VERSION` to 2 in constants.xml.
  - Update existing tests to incorporate the new change to the DS Block header.
- Add new tests for the following components:
  - `UpdateDSComposition`
  - `SaveDSPerformance`
  - `DetermineByzantineNodes`

## Backward Compatibility
<!-- For breaking changes, code must be protected by UPGRADE_TARGET -->
- [ ] This is not a breaking change
- [X] This is a breaking change

The `DirectoryService::VerifyRemovedByzantineNodes` will not pass on old DS Blocks when validating the history.

## Review Suggestion
<!-- How should the reviewers get started on reviewing your pull request-->
<!-- How can the reviewers verify the pull request is working as expected -->
When testing with a local testnet, it should be noted that nodes that are removed for non-performance and invoke the `RejoinAsNormal` call will be stuck because there is no S3 bucket or script to pull the latest state from.

## Status

### Implementation
<!-- Add more TODOs before "ready for review", if any  -->
- [ ] **check that `RejoinAsNormal` works after a node is removed for non-performance.**
- [x] **ready for review**

### Integration Test (Core Team)
<!-- This is for core team only, ignore this if you are a community contributor -->

<!-- append the commit digest to inform the others of the versions you have tested -->
- [x] local machine test
<!-- - [ ] local machine test (commit: bbbbbbbb) -->
<!-- - [ ] local machine test (commit: cccccccc) -->
<!-- - [ ] local machine test (commit: dddddddd) -->
- [x] small-scale cloud test
<!-- - [ ] small-scale cloud test (commit: bbbbbbbb) -->
<!-- - [ ] small-scale cloud test (commit: cccccccc) -->
<!-- - [ ] small-scale cloud test (commit: dddddddd) -->

KaustubhShamshery commented on src/libDirectoryService/DSBlockPostProcessing.cpp:
```diff
@@ -233,38 +229,48 @@ void DirectoryService::UpdateMyDSModeAndConsensusId() {
+       it != m_mediator.m_DSCommittee->end(); ++it) {
+    // Look for my public key.
+    if (m_mediator.m_selfKey.second == it->first) {
+      m_consensusMyID = it - m_mediator.m_DSCommittee->begin();
```
better use the `distance` func

nnamon replied:
Thanks, resolved in https://github.com/Zilliqa/Zilliqa/pull/1587/commits/95906a7d75a2be4474f3fd02b09ad55977828c08.

ShengguangXiao commented on constants.xml:
```diff
@@ -145,6 +145,11 @@
         <COMM_SIZE>200</COMM_SIZE>
         <NUM_DS_ELECTION>5</NUM_DS_ELECTION>
+        <!-- DS Node must meet this minimum threshold to remain in the committee -->
+        <DS_PERFORMANCE_THRESHOLD_PERCENT>0.05</DS_PERFORMANCE_THRESHOLD_PERCENT>
```
DS_PERFORMANCE_THRESHOLD_PERCENT and NUM_DS_BYZANTINE_REMOVED looks never used in the code

ansnunez replied:
@ShengguangXiao DSBlockPreProcessing.cpp

nnamon replied:
Used here: https://github.com/Zilliqa/Zilliqa/pull/1587/files#diff-76372bc310da6631dec75919a39cb4e2R1426

ShengguangXiao commented on src/libDirectoryService/DSBlockPostProcessing.cpp:
```diff
@@ -233,38 +229,48 @@ void DirectoryService::UpdateMyDSModeAndConsensusId() {
-      m_mediator.m_DSCommittee->size()) {
+
+  // Find my new consensus ID.
+  DequeOfNode::iterator it;
```
Put the definition into the for loop, can use auto it = m_mediator.m_DSCommittee->begin(), and use the m_mutexDSCommittee to lock the m_DSCommittee before use it.

nnamon replied:
Thanks, relevant iterators are now using auto instead of explicitly declaring iterator type in this commit: https://github.com/Zilliqa/Zilliqa/pull/1587/commits/83971d8f35d9b5f2eb51d743d8cf3957a46c8e62

ShengguangXiao commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -407,8 +541,14 @@ bool DirectoryService::VerifyPoWOrdering(
+
+  // Count the number of actual PoW winners and removed nodes.
+  std::map<PubKey, Peer> dsPoWWinners =
+      m_pendingDSBlock->GetHeader().GetDSPoWWinners();
```
Use auto

nnamon replied:
Thanks, changed in https://github.com/Zilliqa/Zilliqa/pull/1587/commits/91acff0ca0f2c1d7140c6c7d3eb05d79ec3f12eb.

ShengguangXiao commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -1216,6 +1366,124 @@ bool DirectoryService::ProcessShardingStructure(
+          powDSWinners[it->first] = Peer();
+        } else {
+          powDSWinners[it->first] = it->second;
+        }
```
The under performanced DS node should kick out, if put it into powDSWinner, it will come back again?

ShengguangXiao commented on src/libUtils/DSComposition.cpp:
```diff
@@ -0,0 +1,100 @@
+  for (const auto& DSPowWinner : NewDSMembers) {
+    // Check if the current pow candidate is an existing DS Committee member.
+    // ('loser') and find its index.
+    for (it = dsComm.begin(); it != dsComm.end(); ++it) {
```
use std::find instead of for loop

nnamon replied:
Can't use `std::find` since `dsComm` is a `DequeOfNodes (Pair<PubKey, Peer>)` and `it` may have a zeroed out Peer and `it->first` is a `PubKey`. This causes `no match for ‘operator==’` errors because the type does not match for comparison. `std::find_if` should be used instead but this introduces lambdas which looks even less clear.

nnamon replied:
I made the change to `std::find` initially but reverted it because it breaks the build.

ShengguangXiao replied:
I see

nnamon commented:
Added the unit tests for libUtils/DSComposition.cpp.

ansnunez commented on src/libUtils/DSComposition.h:
```diff
@@ -0,0 +1,32 @@
+#ifndef __DSCOMPOSITION_H__
+#define __DSCOMPOSITION_H__
+
+#include "Logger.h"
```
Minor comments (no need to change for now)
1. I prefer just including the minimum required headers in a header file.  Logger, for example, is not needed here
2. We can probably keep this function inside libDirectoryService instead of libUtils?

nnamon replied:
1. I've removed the unnecessary '#include "Logger.h"', thanks.
2. I can move it to libDirectoryService. The original rationale was that it was split out because libNode had an exact same copy of the code in it's own DSBlockProcessing.cpp in the current codebase and I did not want to introduce even more coupling. I moved the file to libDirectoryService.

ansnunez commented on src/libDirectoryService/DSBlockPostProcessing.cpp:
```diff
@@ -460,18 +446,30 @@ void DirectoryService::StartFirstTxEpoch() {
     if (!found) {
       LOG_EPOCH(WARNING, m_mediator.m_currentEpochNum,
-                "WARNING: Oldest DS node not in any of the new shards!");
+                "WARNING: My DS node is not in any of the new shards! I must "
```
Please change to:
`LOG_EPOCH(WARNING, m_mediator.m_currentEpochNum, "My DS node signed insufficient blocks. Kicked out and invoking RejoinAsNormal now.");`

nnamon replied:
Thanks for the clearer message suggestion. It has been changed.

ansnunez commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -209,27 +209,90 @@ void DirectoryService::ComputeSharding(const VectorOfPoWSoln& sortedPoWSolns) {
     return;
   }
 
+  if (numOfRemovedMembers > powDSWinners.size()) {
```
Is this check necessary?

nnamon replied:
Not really, but it helps with being paranoid and not making so many assumptions about the incoming data.

nnamon replied:
Removed the unnecessary checks with the newly added field in the latest commit.

nnamon commented:
Will be making changes to incorporate @ShengguangXiao's suggestion to introduce a new field in the DSBlock instead of using powWinners.

nnamon commented:
Updated the PR to include the bare minimum to start working on the new field. Fixed all the tests and updated the core messaging code/ds block header to handle it. Incremented the `DSBLOCK_VERSION` constant too.

@ShengguangXiao, one question: Should there be anything to document that this field is introduced in DS Block Version 2 or do I just use the release version like in your example?

https://github.com/Zilliqa/Zilliqa/blob/bc0bc4ce725ade1285a0f0438cbac02645518ea7/src/libMessage/ZilliqaMessage.proto#L83

Logic to use the new field is in the next commit(s).

nnamon commented:
Logic implemented. @ShengguangXiao wanted to check if https://github.com/Zilliqa/Zilliqa/pull/1587/files#diff-76372bc310da6631dec75919a39cb4e2R1193 is acceptable to allow the validation to pass on old blocks. It will only verify the `m_removeDSNodePubkeys` on DS block version 2 and up.

nnamon commented:
Ready for review.

nnamon commented:
Ready for review.

bb111189 commented on src/libDirectoryService/DSBlockPostProcessing.cpp:
```diff
@@ -233,38 +230,47 @@ void DirectoryService::UpdateMyDSModeAndConsensusId() {
-      m_mediator.m_DSCommittee->size()) {
+
+  // Find my new consensus ID.
+  bool isDropout = true;
```
To check for Dropout, I think now we can make use of `m_removeDSNodePubkeys`

nnamon replied:
`dropout` is on two possibilities:

1. is a removed node
2. is an expired node.

This covers both possibilities so I think handling them separately is not the best.

nnamon replied:
To clarify, the `m_removeDSNodePubkeys` field contains only removals for non-performance as suggested in the @ShengguangXiao's proposal to add the new field. The expired nodes are calculated in `UpdateDSCommitteeComposition`. I think this is better than putting expired nodes in the field because it allows some distinction deterministically when just looking at the DS block alone.

bb111189 replied:
I see. Then it does make sense. Maybe we can add an inline comment for the possibilities of dropout
```
is a removed node
is an expired node
```

nnamon replied:
Thanks for the suggestion. Included more comments to make it clearer.

bb111189 commented on src/libMessage/ZilliqaMessage.proto:
```diff
@@ -78,8 +78,9 @@ message ProtoDSBlock
-        repeated PowDSWinners dswinners = 10;  // Added in: v1.0, Deprecated in: N/A
+        repeated PowDSWinners dswinners = 10; // Added in: v1.0, Deprecated in: N/A
         optional DSBlockHashSet hash    = 11; // Added in: v1.0, Deprecated in: N/A
+        repeated ByteArray dsremoved = 12; // Added in: v4.5, Deprecated in: N/A
```
Change to `added in: v2.0` since the version is from block version

nnamon replied:
Thanks, made the change in https://github.com/Zilliqa/Zilliqa/pull/1587/commits/5f896af587a3653bb5cc0aac0d518253d0ba6819

bb111189 commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -222,14 +230,41 @@ void DirectoryService::InjectPoWForDSNode(VectorOfPoWSoln& sortedPoWSolns,
     return;
   }
 
+  if (numOfRemovedMembers > numOfProposedDSMembers) {
```
while the previous implementation is `>`, I think it may make more sense it we use `!=` here

nnamon replied:
No, `numOfRemovedMembers` must be strictly less than or equals to `numOfProposedDSMembers` because `numOfRemovedMembers` contains the members to be removed for non-performance only, not expiry.

bb111189 replied:
Ok then this will make sense.

ShengguangXiao commented on src/libDirectoryService/DSBlockPostProcessing.cpp:
```diff
@@ -287,44 +293,20 @@ void DirectoryService::UpdateMyDSModeAndConsensusId() {
 }
 
-void DirectoryService::UpdateDSCommiteeComposition() {
+void DirectoryService::UpdateDSCommitteeComposition() {
```
Yes, it is better to make it protected

ShengguangXiao commented on src/libMessage/ZilliqaMessage.proto:
```diff
@@ -78,8 +78,9 @@ message ProtoDSBlock
-        repeated PowDSWinners dswinners = 10;  // Added in: v1.0, Deprecated in: N/A
+        repeated PowDSWinners dswinners = 10; // Added in: v1.0, Deprecated in: N/A
         optional DSBlockHashSet hash    = 11; // Added in: v1.0, Deprecated in: N/A
+        repeated ByteArray dsremoved = 12; // Added in: v2.0, Deprecated in: N/A
```
Please align the number and the comment

nnamon replied:
Thanks, done.

ShengguangXiao commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -209,10 +212,15 @@ void DirectoryService::ComputeSharding(const VectorOfPoWSoln& sortedPoWSolns) {
-  // gets included in sharding structure
+void DirectoryService::InjectPoWForDSNode(
+    VectorOfPoWSoln& sortedPoWSolns, unsigned int numOfProposedDSMembers,
+    std::vector<PubKey>& removeDSNodePubkeys) {
```
use const infront of std::vector<PubKey>&

nnamon replied:
Thanks, made the change.

ShengguangXiao commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -43,6 +44,8 @@
 using namespace std;
 using namespace boost::multiprecision;
 
+#define REMOVED_FIELD_DSBLOCK_VERSION 2
```
use const uint32_t instead of #define

nnamon replied:
Thanks, made the change.

ShengguangXiao commented on src/libDirectoryService/DirectoryService.h:
```diff
@@ -631,6 +645,20 @@ class DirectoryService : public Executable {
+      std::vector<PubKey>& removeDSNodePubkeys, uint64_t currentEpochNum,
+      unsigned int numOfFinalBlock, double performanceThreshold,
+      unsigned int maxByzantineRemoved, DequeOfNode& dsComm,
+      std::map<PubKey, uint32_t>& dsMemberPerformance);
```
use const for dsMemberPerformance

nnamon replied:
Thanks, done.

ShengguangXiao commented:
@nnamon Can you do a rebase to sync with the latest master, then i can start to do cloud test for this change.

nnamon commented:
@ShengguangXiao the branch is synced to master.

nnamon commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -68,9 +67,6 @@ unsigned int DirectoryService::ComputeDSBlockParameters(
     }
     powDSWinners[submitter.second] = m_allPoWConns[submitter.second];
     dsWinnerPoWs[submitter.second] = m_allDSPoWs[submitter.second];
-    sortedPoWSolns.erase(
```
@ShengguangXiao to re-order the SortPoWSoln calls, I need to remove the dependency of `sortedPoWSolns` in `DirectoryService::ComputeDSBlockParameters`.

nnamon commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -509,6 +503,22 @@ bool DirectoryService::VerifyPoWOrdering(
+  auto sortedPoWSolns =
+      SortPoWSoln(priorityNodePoWs, true, removeDSNodePubkeys.size());
+
+  // Remove the DS solutions from the PoW solutions.
```
Need to check with you on this portion, if it's needed. I just made sure the same computation was done for the Backup nodes like in the Leader.

nnamon commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -744,10 +755,15 @@ VectorOfPoWSoln DirectoryService::SortPoWSoln(const MapOfPubKeyPoW& mapOfPoWs,
-        ShardSizeCalculator::GetTrimmedShardCount(
-            m_mediator.GetShardSize(false), SHARD_SIZE_TOLERANCE_LO,
-            SHARD_SIZE_TOLERANCE_HI, numNodesTotal);
+        std::min(ShardSizeCalculator::GetTrimmedShardCount(
```
Allowance for extra solutions to be left in is calculated here.

nnamon commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -913,12 +928,28 @@ bool DirectoryService::RunConsensusOnDSBlockWhenDSPrimary() {
+  // Sort and trim the PoW solutions.
+  auto sortedPoWSolns = SortPoWSoln(allPoWs, true, numByzantine);
+
+  // Remove the DS solutions from the PoW solutions.
```
I move the removing of DS solutions from the PoW solutions here from `ComputeDSBlockParameters`, after the DSBlock is computed and the sorted PoW solutions are generated.

nnamon commented:
@ShengguangXiao The requested changes for the rolling upgrade has been made. Do you want me to sync to master too?

ShengguangXiao commented:
> @ShengguangXiao The requested changes for the rolling upgrade has been made. Do you want me to sync to master too?

Yes, please sync to latest master then i can test again @nnamon

nnamon commented:
@ShengguangXiao Synced to master. Build should pass now.

ShengguangXiao commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -1071,6 +1217,20 @@ bool DirectoryService::DSBlockValidator(
+  const uint32_t REMOVED_FIELD_DSBLOCK_VERSION = 2;
+  if (m_pendingDSBlock->GetHeader().GetVersion() >=
+          REMOVED_FIELD_DSBLOCK_VERSION &&
+      m_mediator.m_currentEpochNum >= ENABLE_NEW_FEATURE_DS_NUM) {
```
Change to m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum() >=
      ENABLE_NEW_FEATURE_DS_NUM

ShengguangXiao commented on src/libDirectoryService/DSBlockPreProcessing.cpp:
```diff
@@ -797,21 +917,47 @@ bool DirectoryService::RunConsensusOnDSBlockWhenDSPrimary() {
+
+  // Determine the losers from the performance.
+  unsigned int numByzantine = 0;
+  if (m_mediator.m_currentEpochNum >= ENABLE_NEW_FEATURE_DS_NUM) {
```
Change to m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum() >=
      ENABLE_NEW_FEATURE_DS_NUM

ShengguangXiao commented:
@nnamon The PR is almost ready, just need to change if (m_mediator.m_currentEpochNum >= ENABLE_NEW_FEATURE_DS_NUM) to if (m_mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum() >=
      ENABLE_NEW_FEATURE_DS_NUM), because we need to use DS block number to do the check instead of txn block number

nnamon commented:
@ShengguangXiao thanks, change has been made.

ShengguangXiao commented:
> @ShengguangXiao thanks, change has been made.

Good, i think this PR is ready, will merge after the smart contract upgrade is done

ShengguangXiao commented:
@nnamon @ckyang There is already one field UPGRADE_TARGET_DS_NUM in constants.xml reserved for upgrading breaking change, can you remove ENABLE_NEW_FEATURE_DS_NUM and use UPGRADE_TARGET_DS_NUM instead?

nnamon commented:
@ShengguangXiao changes made, thanks.

bb111189 commented:
Tagging the issue that comes with this PR
https://github.com/Zilliqa/Zilliqa/issues/1517
